visions/swift-testing.md

```diff
@@ -0,0 +1,1024 @@
+* **Allow observing test events:** Some use cases require an ability to observe
+  test events—for example, to perform custom reporting or analysis of results. A
+  testing library should offer hooks for event handling.
+
+### Scalability
+
+* **Parallelize execution:** Many tests can be run in parallel to improve
+  execution time, either using multiple threads in a single process or multiple
+  runner processes. A testing library should offer flexible parallelization
+  options for eligible tests, encourage parallelizing whenever possible, and
+  offer granular control over this behavior. It should also leverage Swift’s
+  data race safety features (such as `Sendable` enforcement) to the fullest
+  extent possible to avoid concurrency bugs.
+* **Repeat a test multiple times with different arguments:** Many tests consist
+  of a template with minor variations—for example, invoking a function multiple
+  times with different arguments each time and validating the result of each
+  invocation. A testing library should make this pattern easy to apply, and
+  include detailed reporting so a failure during a single argument is
+  represented clearly.
+* **Behave consistently across platforms:** Any new testing solution should be
+  cross-platform from its inception and support every platform Swift supports. 
+  Its observable behaviors should be as consistent as possible across those
+  platforms, especially for core responsibilities such as discovering and
+  executing tests.
+
+## Design considerations
+
+Several areas deserve close examination when designing a new testing API. Some,
+because they may benefit from language or compiler toolchain enhancements to
+deliver the ideal experience, and others because they have non-obvious
+reasoning or requirements.
+
+### Expectations
+
+Testing libraries typically offer APIs to compare values—for example, to
+confirm that a function returns an expected result—and report a test failure if
+a comparison does not succeed. Depending on the library, these APIs may be
+called “assertions”, “expectations”, “checks”, “requirements”, “matchers“, or
+other names. In this document we refer to them as **expectations**.
+
+For test failures to be actionable, they need to include enough details to
+understand the problem, ideally without a human manually reproducing the
+failure and debugging. The most important details relevant to expectation
+failures are the values being compared or checked and the kind of expectation
+being performed (e.g. equal, not-equal, less-than, is-not-nil, etc.). Also, if
+any error was caught while evaluating an expression passed to an expectation, 
+that should be included.
+
+Beyond the values of evaluated expressions, there are other pieces of
+information that may be useful to capture and include in expectations:
+
+* The **source code location** of the expectation, typically using the format
+  `#file:#line:#column`. This helps test authors jump quickly to the line of
```

Comment:
minor: is it worth saying #fileID? I believe we mostly moved to using that wherever possible? 🤔

Reply:
Agreed! Done, thanks